File: cp_rust/src/bin/cses-2205-gray-code.rs
// Created by Ayush Biswas at 2025/07/10 12:58
// https://cses.fi/problemset/task/2205
use cp_lib::*;

// @code begin
use cpio::*;

sol! {
    fn solution(
        n: usize,
    ) -> Lines<ListOf<'\0', u8>> {
        (0..1<<n).map(|i| gray(i, n).into()).collect()
    }
}

fn gray(i: usize, n: usize) -> Vec<u8> {
    (0..n)
        .rev()
        .map(|j| {
            let a = (i >> (j + 1)) % 2;
            let b = (i >> j) % 2;
            (a ^ b) as u8
        })
        .collect()
}

// @code end

File: cp_haskell/bin/cses-2205-gray-code.hs
-- Created by Ayush Biswas at 2026/03/04 19:30
-- https://cses.fi/problemset/task/2205

-- @code begin

import Data.Bits (xor)
import Text.Printf (printf)

main :: IO ()
main = do
  n :: Int <- read <$> getLine
  let res = (gray . binary n <$> [0 .. 2 ^ n - 1]) <*> [[]]
  mapM_ (\code -> mapM_ (printf "%d") code *> printf "\n") res
  where
    gray xs@(x : _) [] = gray xs [x]
    gray (x : y : xs) as = gray (y : xs) $ (x `xor` y) : as
    gray _ res = res

    binary :: Int -> Int -> [Int]
    binary n x = reverse $ (`mod` 2) <$> take n (iterate (`div` 2) x)

-- @code end